### PR TITLE
[SPARK-12195] [SQL] Adding BigDecimal, Date and Timestamp into Encoder

### PR DESCRIPTION
**Pull request by gatorsmile**
This PR is to add three more data types into Encoder, including `BigDecimal`, `Date` and `Timestamp`. 

@marmbrus @cloud-fan @rxin Could you take a quick look at these three types? Not sure if it can be merged to 1.6. Thank you very much!

### CODE REVIEW

File: sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java [hunk 1]
```diff
@@ -386,6 +389,20 @@ public void testNestedTupleEncoder() {
   }
 
   @Test
+  public void testTypeEncoder() {
```
Comment (cloud-fan) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java]: `testPrimitiveEncoder`?

Comment (gatorsmile) [inline sql/core/src/test/java/test/org/apache/spark/sql/JavaDatasetSuite.java]: Sure. Thank you! Let me change it now.

**Comment by cloud-fan** [comment]
LGTM

**Comment by marmbrus** [comment]
Thanks, merging to master and 1.6

**Pull request closed without merging**